### PR TITLE
Core: rnotwhite -> rhtmlnotwhite and jQuery.trim -> stripAndCollapse

### PR DESCRIPTION
**Pull request by timmywil**
- Renames and changes rnotwhite to focus on HTML whitespace chars
- Change internal use of jQuery.trim to more accurate strip and collapse
- Adds tests to ensure HTML space characters are retained where valid
- Doesn't add tests where the difference is inconsequential and
  existing tests are adequate.

Fixes gh-3003
Fixes gh-3072

* [x] All authors have signed the CLA at https://contribute.jquery.com/CLA/
* [x] New tests have been added to show the fix or feature works
* [x] Grunt build and unit tests pass locally with these changes
* [x] ~~If needed, a docs issue/PR was created at https://github.com/jquery/api.jquery.com~~

### CODE REVIEW

File: src/core/stripAndCollapse.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+define( function() {
+	"use strict";
+
+	// Strip and collapse whitespace according to HTML spec
+	// https://html.spec.whatwg.org/multipage/infrastructure.html#strip-and-collapse-whitespace
+	var rhtmlSpace = /[\x20\t\r\n\f]+/g,
```
Comment (mgol) [inline src/core/stripAndCollapse.js]: Can't you write a regular space (` `) instead of the hex code (`\x20`)?

File: test/unit/ajax.js [hunk 2]
```diff
@@ -2327,6 +2327,15 @@ if ( typeof window.ArrayBuffer === "undefined" || typeof new XMLHttpRequest().re
 		} );
 	} );
 
+	// Selector should be trimmed to avoid leading spaces (#14773)
+	// Selector should include any valid non-HTML whitespace (#3003)
+	QUnit.asyncTest( "jQuery.fn.load( URL_SELECTOR with non-HTML whitespace(#3003) )", 1, function( assert ) {
+		jQuery( "#first" ).load( "data/test3.html   #whitespace\xA0 ", function() {
+			assert.strictEqual( jQuery( this ).children( "div" ).length, 1, "Verify that specific elements were injected" );
+			QUnit.start();
```
Comment (markelog) [inline test/unit/ajax.js]: See https://api.qunitjs.com/QUnit.start/

File: src/var/rnothtmlwhite.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+define( function() {
+	"use strict";
+
+	// Only count HTML whitespace
+	// Other whitespace should count in values
+	// https://html.spec.whatwg.org/multipage/infrastructure.html#space-character
+	return ( /[^\x20\t\r\n\f]+/g );
```
Comment (mgol) [inline src/var/rnothtmlwhite.js]: Same remark as above.

File: test/unit/ajax.js (hunk 2 above)
Comment (markelog) [inline test/unit/ajax.js]: Also, https://github.com/platinumazure/eslint-plugin-qunit, not sure if it would catch it, but still, what do you think?

File: test/unit/attributes.js [hunk 4]
```diff
@@ -657,6 +657,28 @@ QUnit.test( "removeAttr(Multi String, variable space width)", function( assert )
...
+		};
+
+	jQuery.each( tests, function( key, val ) {
+		assert.equal( div.attr( key ), val, "Attribute `" + key + "` exists, and has a value of `" + val + "`" );
+	} );
+
+	div.removeAttr( "id   data-\xA0 title  rel  " );
+
+	jQuery.each( tests, function( key ) {
+		assert.equal( div.attr( key ), undefined, "Attribute `" + key + "` was removed" );
```
Comment (markelog) [inline test/unit/attributes.js]: Do we want to start using backticks in strings?

File: test/unit/ajax.js (hunk 2 above)
Comment (mgol) [inline test/unit/ajax.js]: We should not use `QUnit.asyncTest` for new tests at all, should we?

Comment (mgol) [inline test/unit/ajax.js]: @markelog I wanted to look at the docs but I hit this: https://github.com/platinumazure/eslint-plugin-qunit/issues/53. :)

File: test/unit/attributes.js [hunk 5]
```diff
@@ -1117,7 +1139,7 @@ QUnit.test( "val(select) after form.reset() (Bug #2551)", function( assert ) {
 } );
 
 QUnit.test( "select.val(space characters) (gh-2978)", function( assert ) {
-	assert.expect( 35 );
+	assert.expect( 37 );
```
Comment (markelog) [inline test/unit/attributes.js]: Maybe we could move this to separate test?

File: test/unit/attributes.js [hunk 6]
```diff
@@ -657,6 +657,28 @@ QUnit.test( "removeAttr(Multi String, variable space width)", function( assert )
 	} );
 } );
 
+QUnit.test( "removeAttr(Multi String, non-HTML whitespace is valid in attribute names (gh-3003)", function( assert ) {
+	assert.expect( 8 );
+
+	var div = jQuery( "<div id='a' data-\xA0='b' title='c' rel='d'></div>" ),
+		tests = {
```
Comment (markelog) [inline test/unit/attributes.js]: Just fyi, you can use separate `var` for this

File: test/unit/ajax.js (hunk 2 above)
Comment (markelog) [inline test/unit/ajax.js]: Yeah, https://github.com/platinumazure/eslint-plugin-qunit/tree/master/docs/rules those are available though, you should send them a PR :)

Comment (markelog) [inline test/unit/ajax.js]: > We should not use QUnit.asyncTest for new tests at all, should we?

Yeah, that's what I meant by my first comment :)

File: src/core/stripAndCollapse.js (hunk 1 above)
Comment (markelog) [inline src/core/stripAndCollapse.js]: This is somehow different from `[ \t\r\n\v\f]` -> `\s` right? You guys already talked about this I believe?

File: test/unit/ajax.js (hunk 2 above)
Comment (platinumazure) [inline test/unit/ajax.js]: Apologies for the broken links. Rule `no-async-test` should definitely catch this, let me know if I can help with integration.

File: src/core/stripAndCollapse.js (hunk 1 above)
Comment (timmywil) [inline src/core/stripAndCollapse.js]: > This is somehow should be different from [ \t\r\n\v\f] -> \s right?

Right

File: test/unit/attributes.js (hunk 4 above)
Comment (timmywil) [inline test/unit/attributes.js]: Not sure what you mean. You mean use es2015 template strings? That wouldn't work in all of our environments.

File: test/unit/attributes.js (hunk 5 above)
Comment (timmywil) [inline test/unit/attributes.js]: This is what this test was meant for originally.

File: test/unit/attributes.js (hunk 6 above)
Comment (timmywil) [inline test/unit/attributes.js]: This is just a copy/paste based on another test, but I can do that.

File: src/core/stripAndCollapse.js (hunk 1 above)
Comment (markelog) [inline src/core/stripAndCollapse.js]: Could you give a me a context?

Comment (timmywil) [inline src/core/stripAndCollapse.js]: > Can't you write a regular space () instead of the hex code (\x20)?

We could, but I like that this is more apparent/explicit.

Comment (dmethvin) [inline src/core/stripAndCollapse.js]: Also it might be a tab. I prefer the `\x20` for that reason.

Comment (mgol) [inline src/core/stripAndCollapse.js]: It's more characters and a space is a space. :) But I see we're using it in other places like that already. Changing all instances to a space saves only one byte gzipped.

Comment (markelog) [inline src/core/stripAndCollapse.js]: > Also it might be a tab.

Not sure to what you referring to.

For the reference canonical way is to use ` `, don't have strong opinion about it though

Comment (mgol) [inline src/core/stripAndCollapse.js]: @dmethvin No, `/ /` doesn't match a tab.

Comment (markelog) [inline src/core/stripAndCollapse.js]: `\x20` for most though, i think, that symbol doesn't immediately associated with ` `

File: test/unit/attributes.js (hunk 6 above)
Comment (markelog) [inline test/unit/attributes.js]: I'm saying output will be 

> Attribute `id` exists, and has a value of `a`

even though we usually do something like this

> Attribute "id" exists, and has a value of "a"

File: test/unit/attributes.js (hunk 4 above)
Comment (markelog) [inline test/unit/attributes.js]: I'm saying output will be 

> Attribute `id` exists, and has a value of `a`

even though we usually do something like this

> Attribute "id" exists, and has a value of "a"

File: src/core/stripAndCollapse.js (hunk 1 above)
Comment (dmethvin) [inline src/core/stripAndCollapse.js]: Right, in the source it could be a tab. Unlikely, but the `\x20

File: test/unit/attributes.js (hunk 5 above)
Comment (markelog) [inline test/unit/attributes.js]: Yeah, sure, i'm saying it would be better not to change the tests, but add different ones. If we would use the terminology, previously that was an unit-test, now we found a bug, so it should be a regression test, so it might make sense not to unite those

File: src/core/stripAndCollapse.js (hunk 1 above)
Comment (markelog) [inline src/core/stripAndCollapse.js]: ooh, i see

File: test/unit/attributes.js (hunk 4 above)
Comment (timmywil) [inline test/unit/attributes.js]: oh, I see. Sure!

File: test/unit/attributes.js (hunk 5 above)
Comment (timmywil) [inline test/unit/attributes.js]: The "2 added tests" is a misleading. These tests actually needed editing, not additions.

Comment (markelog) [inline test/unit/attributes.js]: Oh, so these editions are unrelated to bug fix? I would move them to different commit, but not feeling strong about this

Comment (timmywil) [inline test/unit/attributes.js]: They're related. It's just that there aren't really any new tests to split off in this particular case.

Comment (timmywil) [inline test/unit/attributes.js]: To be specific, the bug fix cause these tests to fail. Just editing them accomplished the same thing as adding another test to verify the correct behavior. We were verifying buggy behavior before.

File: src/core/stripAndCollapse.js (hunk 1 above)
Comment (markelog) [inline src/core/stripAndCollapse.js]: Found the context, LGTM
